src/Servers/Kestrel/test/InMemory.FunctionalTests/RequestTests.cs

```diff
@@ -253,6 +253,68 @@ public async Task CanHandleRequestsWithUrlInAbsoluteForm(string requestUrl, stri
             }
         }
 
+        [Fact]
+        public async Task ExecutionContextDoesNotLeakAcrossRequestsOnSameConnection()
+        {
+            var local = new AsyncLocal<int>();
+
+            // It's important this method isn't async as that will revert the ExecutionContext
+            Task ExecuteApplication(HttpContext context)
+            {
+                var value = local.Value;
+                Assert.Equal(0, value);
+                local.Value++;
+
+                context.Response.OnStarting(() =>
+                {
+                    local.Value++;
+                    return Task.CompletedTask;
+                });
+
+                context.Response.OnCompleted(() =>
+                {
+                    local.Value++;
+                    return Task.CompletedTask;
+                });
+
+                context.Response.ContentLength = 1;
+                return context.Response.WriteAsync(value.ToString());
+            }
+
+            var testContext = new TestServiceContext(LoggerFactory);
+            await using (var server = new TestServer(ExecuteApplication, testContext))
+            {
+                using (var connection = server.CreateConnection())
+                {
+                    await connection.Send(
+                        "GET / HTTP/1.1",
+                        "Host:",
+                        "",
+                        "");
+
+                    await connection.Receive(
+                        "HTTP/1.1 200 OK",
+                        $"Date: {testContext.DateHeaderValue}",
+                        "Content-Length: 1",
+                        "",
+                        "0");
+
+                    await connection.Send(
+                        "GET / HTTP/1.1",
+                        "Host:",
+                        "",
+                        "");
+
+                    await connection.Receive(
+                        "HTTP/1.1 200 OK",
+                        $"Date: {testContext.DateHeaderValue}",
+                        "Content-Length: 1",
+                        "",
+                        "0");
+                }
+            }
+        }
```

Comment:
The following the HTTP/2 version of the test to catch possible regression. It should be able to go into any Http2TestBase-derived test class.

```C#
[Fact]
public async Task Frame_MultipleStreams_DoesNotLeakAcrossRequests()
{
    var local = new AsyncLocal<int>();

    // It's important this method isn't async as that will revert the ExecutionContext
    Task ExecuteApplication(HttpContext context)
    {
        var value = local.Value;
        Assert.Equal(0, value);
        local.Value++;

        context.Response.OnStarting(() =>
        {
            Assert.Equal(1, local.Value);
            local.Value++;
            return Task.CompletedTask;
        });

        context.Response.OnCompleted(() =>
        {
            Assert.Equal(0, local.Value);
            local.Value++;
            return Task.CompletedTask;
        });

        context.Response.ContentLength = 1;
        return context.Response.WriteAsync(value.ToString());
    }

    await InitializeConnectionAsync(ExecuteApplication);

    await StartStreamAsync(1, _browserRequestHeaders, endStream: true);
    await ExpectAsync(Http2FrameType.HEADERS,
        withLength: 55,
        withFlags: (byte)Http2HeadersFrameFlags.END_HEADERS,
        withStreamId: 1);

    var dataFrame1 = await ExpectAsync(Http2FrameType.DATA,
        withLength: 1,
        withFlags: (byte)Http2DataFrameFlags.NONE,
        withStreamId: 1);

    await ExpectAsync(Http2FrameType.DATA,
        withLength: 0,
        withFlags: (byte)Http2DataFrameFlags.END_STREAM,
        withStreamId: 1);

    Assert.Equal((byte)'0', dataFrame1.Payload.Span[0]);

    await StartStreamAsync(3, _browserRequestHeaders, endStream: true);
    await ExpectAsync(Http2FrameType.HEADERS,
        withLength: 55,
        withFlags: (byte)Http2HeadersFrameFlags.END_HEADERS,
        withStreamId: 3);

    var dataFrame2 = await ExpectAsync(Http2FrameType.DATA,
        withLength: 1,
        withFlags: (byte)Http2DataFrameFlags.NONE,
        withStreamId: 3);

    await ExpectAsync(Http2FrameType.DATA,
        withLength: 0,
        withFlags: (byte)Http2DataFrameFlags.END_STREAM,
        withStreamId: 3);

    Assert.Equal((byte)'0', dataFrame2.Payload.Span[0]);

    await StopConnectionAsync(expectedLastStreamId: 3, ignoreNonGoAwayFrames: false);
}
```